### DISCUSSION TITLE
Allow using static SomeClass to bring overloaded operators into scope.

### DISCUSSION THREAD

**Discussion by lostmsu** [dotnet/csharplang#2294] [Language Ideas] [answered]
This proposal is to enable adding operators to existing 3rd party types, or having different implementations of the same operators (for example, separate sets for GPU/CPU computation).

Proposal consists of two changes:
1. Enable defining operators for arbitrary types inside static classes. E.g.
```csharp
static class ListMath {
  public static List operator+ (List a, List b)
    => new List(a.Count + b.Count).AddRange(a).AddRange(b);
}
```

2. Enable bringing them into scope via `using static`:
```csharp
using static ListMath;

var a = new List{ 1 };
var b = new List{ 2 };
var c = a + b;
```

3. (optional, but highly desired) Allow operators in 1 to be generic:
```csharp
public static List<T> operator+<T> (List<T> a, List<T> b) => new List(a.Count + b.Count).AddRange(a).AddRange(b);
```

Resolution of these operators should follow resolution rules of extension methods. E.g. if either operand has matching native operator defined, use that, otherwise use this extension operator.

**Comment by Unknown6656** [comment]
related: https://github.com/dotnet/roslyn/issues/11159, https://github.com/dotnet/roslyn/issues/6136, https://github.com/dotnet/roslyn/issues/5165, https://github.com/dotnet/roslyn/issues/4945, https://github.com/dotnet/roslyn/issues/5624 (afaik some of the referenced proposals would allow extension operators, which should be what you request)

**Comment by lostmsu** [comment, marked as answer]
@Unknown6656 that was moved to https://github.com/dotnet/csharplang/issues/192

**Comment by Unknown6656** [comment]
Ah thanks, I was unable to find it with a superficial search.
